### PR TITLE
CGU-5184 Add std::iostream version of model loading

### PR DESCRIPTION
**Pull request by tgrobler**

### CODE REVIEW

File: include/caffe/net.hpp [hunk 1]
```diff
@@ -26,6 +27,8 @@ class Net {
   explicit Net(const NetParameter& param);
   explicit Net(const string& param_file, Phase phase,
       const int level = 0, const vector<string>* stages = NULL);
+  explicit Net(std::istream& param_stream, const string& name_to_log, Phase phase,
+    const int level = 0, const vector<string>* stages = NULL);
```
Comment (ultinous-nil) [inline include/caffe/net.hpp:31]: prefix comma, postfix const

File: include/caffe/net.hpp [hunk 2]
```diff
@@ -26,6 +27,8 @@ class Net {
   explicit Net(const NetParameter& param);
   explicit Net(const string& param_file, Phase phase,
       const int level = 0, const vector<string>* stages = NULL);
```
Comment (ultinous-nil) [inline include/caffe/net.hpp:29]: prefix comma, postfix const

File: include/caffe/net.hpp [hunk 3]
```diff
@@ -113,6 +116,7 @@ class Net {
   void CopyTrainedLayersFrom(const NetParameter& param);
   void CopyTrainedLayersFrom(const string trained_filename);
   void CopyTrainedLayersFromBinaryProto(const string trained_filename);
+  void CopyTrainedLayersFromBinaryProto(std::istream& trained_stream, const string& name_to_log);
```
Comment (ultinous-nil) [inline include/caffe/net.hpp:119]: postfix const.

File: include/caffe/util/upgrade_proto.hpp [hunk 4]
```diff
@@ -16,8 +16,12 @@ bool UpgradeNetAsNeeded(const string& param_file, NetParameter* param);
 // Read parameters from a file into a NetParameter proto message.
 void ReadNetParamsFromTextFileOrDie(const string& param_file,
                                     NetParameter* param);
+void ReadNetParamsFromTextStreamOrDie(std::istream& param_stream, const string& name_to_log,
```
Comment (ultinous-nil) [inline include/caffe/util/upgrade_proto.hpp:19]: & to the right

File: include/caffe/util/upgrade_proto.hpp [hunk 5]
```diff
@@ -16,8 +16,12 @@ bool UpgradeNetAsNeeded(const string& param_file, NetParameter* param);
...
+                                    NetParameter* param);
 void ReadNetParamsFromBinaryFileOrDie(const string& param_file,
                                       NetParameter* param);
+void ReadNetParamsFromBinaryStreamOrDie(std::istream& param_stream, const string& name_to_log,
+  NetParameter* param);
```
Comment (ultinous-nil) [inline include/caffe/util/upgrade_proto.hpp:24]: ...

File: src/caffe/net.cpp [hunk 6]
```diff
@@ -40,6 +40,22 @@ Net<Dtype>::Net(const string& param_file, Phase phase,
...
 }
 
+template <typename Dtype>
+Net<Dtype>::Net(std::istream& param_stream, const string& name_to_log, Phase phase,
+  const int level, const vector<string>* stages) {
+  NetParameter param;
+  ReadNetParamsFromTextStreamOrDie(param_stream, name_to_log, &param);
+  // Set phase, stages and level
+  param.mutable_state()->set_phase(phase);
+  if (stages != NULL) {
```
Comment (ultinous-nil) [inline src/caffe/net.cpp]: Why not if(stages)? Why not nullptr? NULL is obsolete.

File: src/caffe/util/io.cpp [hunk 7]
```diff
@@ -45,6 +47,13 @@ bool ReadProtoFromTextFile(const char* filename, Message* proto) {
   return success;
 }
 
+bool ReadProtoFromTextStream(std::istream& stream, Message* proto) {
+  IstreamInputStream* input = new IstreamInputStream(&stream);
```
Comment (ultinous-nil) [inline src/caffe/util/io.cpp]: Why "new"? Not exception safe. Unnecessary heap allocation.

File: src/caffe/util/io.cpp [hunk 8]
```diff
@@ -53,6 +62,12 @@ void WriteProtoToTextFile(const Message& proto, const char* filename) {
   close(fd);
 }
 
+void WriteProtoToTextStream(const Message& proto, std::ostream& stream) {
+  OstreamOutputStream* output = new OstreamOutputStream(&stream);
+  CHECK(google::protobuf::TextFormat::Print(proto, output));
```
Comment (ultinous-nil) [inline src/caffe/util/io.cpp]: Why new (again)?

File: src/caffe/util/io.cpp [hunk 9]
```diff
@@ -72,6 +87,18 @@ bool ReadProtoFromBinaryFile(const char* filename, Message* proto) {
   return success;
 }
 
+bool ReadProtoFromBinaryStream(std::istream& stream, Message* proto) {
+  IstreamInputStream* raw_input = new IstreamInputStream(&stream);
```
Comment (ultinous-nil) [inline src/caffe/util/io.cpp]: Why new?

File: src/caffe/util/upgrade_proto.cpp [hunk 10]
```diff
@@ -92,13 +92,27 @@ void ReadNetParamsFromTextFileOrDie(const string& param_file,
...
 
+void ReadNetParamsFromTextStreamOrDie(std::istream& param_stream, const string& name_to_log,
+  NetParameter* param) {
+  CHECK(ReadProtoFromTextStream(param_stream, param))
+    << "Failed to parse NetParameter stream: " << name_to_log;
+  UpgradeNetAsNeeded(name_to_log, param);
+}
+
 void ReadNetParamsFromBinaryFileOrDie(const string& param_file,
                                       NetParameter* param) {
```
Comment (ultinous-nil) [inline src/caffe/util/upgrade_proto.cpp:103]: Horizontal alignment is not OK. Only logical indentation is ok

**Review by ultinous-nil — CHANGES_REQUESTED** [review]

File: include/caffe/net.hpp (hunk 2 above)
Comment (ultinous-jlocki) [reply include/caffe/net.hpp:29]: > prefix comma, postfix const

This is just an extension to caffe code. We have to use caffe coding style.

File: include/caffe/net.hpp (hunk 1 above)
Comment (ultinous-jlocki) [reply include/caffe/net.hpp:31]: > prefix comma, postfix const

This is just an extension to caffe code. We have to use caffe coding style.

File: include/caffe/net.hpp (hunk 3 above)
Comment (ultinous-jlocki) [reply include/caffe/net.hpp:119]: > postfix const.

This is just an extension to caffe code. We have to use caffe coding style.

File: include/caffe/util/upgrade_proto.hpp (hunk 4 above)
Comment (ultinous-jlocki) [reply include/caffe/util/upgrade_proto.hpp:19]: > & to the right

This is just an extension to caffe code. We have to use caffe coding style.

File: include/caffe/util/upgrade_proto.hpp (hunk 5 above)
Comment (ultinous-jlocki) [reply include/caffe/util/upgrade_proto.hpp:24]: > ...

....

**Review by ultinous-jlocki — COMMENTED** [review]
Coding style has to be the same as the original of Caffe. Only remove the unnecessary heap allocations.

File: src/caffe/net.cpp (hunk 6 above)
Comment (tgrobler) [reply src/caffe/net.cpp]: fixed

File: src/caffe/util/io.cpp (hunk 7 above)
Comment (tgrobler) [reply src/caffe/util/io.cpp]: fixed

File: src/caffe/util/io.cpp (hunk 8 above)
Comment (tgrobler) [reply src/caffe/util/io.cpp]: fixed

File: src/caffe/util/io.cpp (hunk 9 above)
Comment (tgrobler) [reply src/caffe/util/io.cpp]: fixed

**Comment by tgrobler** [comment]
Fixed version pushed.

**Review by ultinous-nil — APPROVED** [review]

**Pull request merged**